src/transformers/testing_utils.py

```diff
@@ -1615,3 +1616,38 @@ def new_request(self, method, **kwargs):
             self.other_request_count += 1
 
         return self.old_request(method=method, **kwargs)
+
+
+# adapted from https://stackoverflow.com/questions/32163436/python-decorator-for-printing-every-line-executed-by-a-function
+class set_reproducible(ContextDecorator):
+    r"""
+    A context decorator to set a manual seed before each PyTorch's forward pass. This is useful for models that
+    involves stochasticity during their forward pass to ensure reproducibility and the capability to run som tests.
+    """
+
+    def __init__(self, seed=42):
+        super().__init__()
+        self.seed = seed
+
+    def __enter__(self):
+        sys.settrace(self.trace_calls)
+
+    def __exit__(self, *exc):
+        # stop tracing functions after quitting
+        # the context manager
+        sys.settrace(None)
+
+    def trace_calls(self, frame, event, arg):
+        # We want to only trace our call events
+        # and setting the seed before the forward function
+        if event != "call":
+            return
+        elif frame.f_code.co_name != "forward":
+            return
+        # Set the seed when it is a call to a forward function
+        return self.set_seed
+
+    def set_seed(self, frame, event, arg):
+        # Set the seed
+        torch.manual_seed(self.seed)
```

Comment:
Can we maybe use the `set_seed` function here from our utils (so it sets all seeds and not just PyTorch, while also working when PyTorch is not installed)?

Reply:
It seems that yes we can! added the proposed change in cb7a4029794427dfeb78f1c4cf50a277f709289d